packages/shared/src/hooks/vote/utils.ts

```diff
@@ -0,0 +1,129 @@
+import { QueryClient } from 'react-query';
+import { Edge } from '../../graphql/common';
+import {
+  generateQueryKey,
+  RequestKey,
+  updateReadingHistoryListPost,
+} from '../../lib/query';
+import { LoggedUser } from '../../lib/user';
+import { ReadHistoryInfiniteData } from '../useInfiniteReadingHistory';
+import {
+  UseVotePostMutationProps,
+  UseVotePostProps,
+  voteMutationHandlers,
+} from './types';
+import { PostItem } from '../../graphql/posts';
+import { optimisticPostUpdateInFeed } from '../../lib/feed';
+import { FeedItem, UpdateFeedPost } from '../useFeed';
+
+export const mutateVoteReadHistoryPost = ({
+  id,
+  vote,
+  data,
+  user,
+  queryClient,
+}: UseVotePostMutationProps & {
+  data: ReadHistoryInfiniteData;
+  user: LoggedUser;
+  queryClient: QueryClient;
+}): ReturnType<UseVotePostProps['onMutate']> => {
+  const mutationHandler = voteMutationHandlers[vote];
+
+  if (!mutationHandler) {
+    return undefined;
+  }
+
+  const findPost = (edge: Edge<PostItem>) => edge.node.post.id === id;
+
+  const postPageIndex = data.pages.findIndex((page) =>
+    page.readHistory.edges.some(findPost),
+  );
+
+  if (postPageIndex === -1) {
+    return undefined;
+  }
+
+  const postIndexToUpdate =
+    data.pages[postPageIndex].readHistory.edges.findIndex(findPost);
+
+  if (postIndexToUpdate === -1) {
+    return undefined;
+  }
+
+  const readingHistoryQueryKey = generateQueryKey(
+    RequestKey.ReadingHistory,
+    user,
+  );
+
+  return updateReadingHistoryListPost({
+    queryKey: readingHistoryQueryKey,
+    queryClient,
+    pageIndex: postPageIndex,
+    index: postIndexToUpdate,
+    manipulate: (currentPost) => {
+      return {
+        ...currentPost,
+        ...mutationHandler(currentPost),
+      };
+    },
+  });
+};
+
+export const mutateVoteFeedPost = ({
+  id,
+  vote,
+  items,
+  updatePost,
+}: UseVotePostMutationProps & {
+  items: FeedItem[];
+  updatePost: UpdateFeedPost;
+}): ReturnType<UseVotePostProps['onMutate']> => {
+  if (!items) {
+    return undefined;
+  }
+
+  const mutationHandler = voteMutationHandlers[vote];
+
+  if (!mutationHandler) {
+    return undefined;
+  }
+
+  const postIndexToUpdate = items.findIndex(
+    (item) => item.type === 'post' && item.post.id === id,
```

Comment:
Not sure, but don't we have a type for 'post'

Reply:
Not sure what you mean?

Reply:
I think this is for FeedItem which has a type: `ad`, `placeholder`, and a `post`.

Reply:
I mean a enum typed version of `post` for this context

Reply:
Nope, it's typed as string, see `PostItem` in `useFeed`

Reply:
Ah ok, don't mind my comment then it's not urgent :)